src/com/amazon/ionelement/api/StructUpdateBuilder.kt

```diff
@@ -0,0 +1,121 @@
+package com.amazon.ionelement.api
+
+import com.amazon.ionelement.api.StructUpdateBuilder.FieldUpdateMode.*
+import com.amazon.ionelement.impl.*
+import kotlinx.collections.immutable.toPersistentList
+import kotlinx.collections.immutable.toPersistentMap
+
+/**
+ * An [IonPath] represents a traversal into a nested IonElement made of [StructElement]s and [SeqElement]s
+ */
+public class IonPath(public val elements: List<IonPathElement>) : Collection<IonPathElement> by elements {
+    public constructor(vararg elements: IonPathElement) : this(elements.asList())
+
+    public fun tail(): IonPath {
+        return IonPath(elements.drop(1))
+    }
+}
+
+public sealed class IonPathElement
+
+/**
+ * Traverse into a nested [StructElement] by its field name.
+ */
+public data class Field(val fieldName: String) : IonPathElement()
+
+/**
+ * Traverse into a nested [SeqElement] by its numerical index.
+ */
+public data class Index(val index: Int) : IonPathElement()
+
+/**
+ * Fetch an element from a nested IonElement using the [path] to traverse struct fields and [SeqElement] indices
+ */
+public tailrec fun IonElement.getPath(path: IonPath): IonElement {
+    return when (path.isEmpty()) {
+        true -> this
+        else -> when (val element = path.first()) {
+            is Field -> this.asAnyElement().asStruct()[element.fieldName].getPath(path.tail())
+            is Index -> this.asAnyElement().asSeq().values[element.index].getPath(path.tail())
+        }
+    }
+}
+
+/**
+ * Implement [] operator for fetching a nested element
+ */
+public operator fun IonElement.get(path: IonPath): IonElement {
+    return this.getPath(path)
+}
+
+/**
+ * A [StructUpdateBuilder] builder implements a typesafe builder that allows updating a [StructElement] by adding
+ * and replacing existing [StructField]s similar to the Kotlin builtin `copy` function for data classes.
+ */
+public class StructUpdateBuilder(
+    private val sourceStruct: StructElement,
+    private val defaultMode: FieldUpdateMode
+) {
+    private val fieldUpdates: MutableMap<String, MutableList<AnyElement>> =
+        sourceStruct.fields.groupBy({ it.name }, { it.value })
+            .mapValues { it.value.toMutableList() }
+            .toMutableMap()
```

Comment:
>  I would want to make sure that within a group of operations, only single copy is made with all of the changes, instead of a copy for each change, which the current builder ensures.

I can elaborate a bit more. The implementation of `MutableStructFields` could use a `MutableMap` or a `MutableList` behind the scenes, so we would create one shallow copy when we create the `MutableStructFieldsImpl` instance, and then one shallow copy when we pass the instance to the constructor or `copy()` function.

> I was also thinking, that since StructElement and SeqElement already use PersistentMap and PersistentList we should be able to expose those to leverage copy-on-write semantics, again as long as multiple updates only create a single copy. I just wasn't sure how receptive you would be to changing the internals too much.

I'm open to changing the internals, but I'm hesitant to expose the persistent collections because that has the potential to limit our ability to change implementation details in the future. We would want to be really sure before we commit to changing the persistent collections from being an implementation detail to being part of the API.

Reply:
It's been a while, but after coming back, I really like the `mutableFields` approach. The interface exposes the data in a writeable way and doesn't leak the internals. Is someone currently working on implementing this?

Reply:
Nope, no one is implementing it right now. Since it's a fairly thin layer built on top of `ion-java`, generally `ion-element-kotlin` does not get feature development unless someone requests/contributes a feature.